system/security/securesocket/securesocket.cpp

```diff
@@ -1177,105 +1284,123 @@ class CSecureSocketContext : implements ISecureSocketContext, public CInterface
         SSL_CTX_set_session_id_context(m_ctx, (const unsigned char*)"hpccsystems", 11);
     }
 
-public:
-    IMPLEMENT_IINTERFACE;
-    CSecureSocketContext(SecureSocketType sockettype)
+    void initContext(SecureSocketType sockettype)
     {
-        m_verify = false;
-        m_address_match = false;
-
         if(sockettype == ClientSocket)
             m_meth = SSLv23_client_method();
         else
             m_meth = SSLv23_server_method();
 
-        m_ctx = SSL_CTX_new(m_meth);
+        m_ctx.setown(SSL_CTX_new(m_meth));
 
         if(!m_ctx)
-        {
-            throw MakeStringException(-1, "ctx can't be created");
-        }
+            throw makeStringExceptionV(-1, "ctx can't be created");
 
         if (sockettype == ServerSocket)
             setSessionIdContext();
-
-        SSL_CTX_set_mode(m_ctx, SSL_CTX_get_mode(m_ctx) | SSL_MODE_AUTO_RETRY);
     }
 
-    CSecureSocketContext(const char* certfile, const char* privkeyfile, const char* passphrase, SecureSocketType sockettype)
+    void setCertificate(const char *certFileorBuf)
     {
-        m_verify = false;
-        m_address_match = false;
-
-        if(sockettype == ClientSocket)
-            m_meth = SSLv23_client_method();
-        else
-            m_meth = SSLv23_server_method();
-
-        m_ctx = SSL_CTX_new(m_meth);
+        if (isEmptyString(certFileorBuf))
+            return;
 
-        if(!m_ctx)
+        if (isPEMBuffer(certFileorBuf))
         {
-            throw MakeStringException(-1, "ctx can't be created");
+            // can have multiple certs in buffer
+            if (!useCertificateChainPEMBuffer(m_ctx, certFileorBuf))
+            {
+                char errbuf[512];
+                ERR_error_string_n(ERR_get_error(), errbuf, 512);
+                throw makeStringExceptionV(-1, "error loading certificate chain %s", errbuf);
+            }
         }
-
-        if (sockettype == ServerSocket)
-            setSessionIdContext();
-
-        password.set(passphrase);
-        SSL_CTX_set_default_passwd_cb_userdata(m_ctx, (void*)password.str());
-        SSL_CTX_set_default_passwd_cb(m_ctx, pem_passwd_cb);
-
-        if (SSL_CTX_use_certificate_chain_file(m_ctx, certfile)<=0)
+        else if (SSL_CTX_use_certificate_chain_file(m_ctx, certFileorBuf) <= 0)
         {
             char errbuf[512];
             ERR_error_string_n(ERR_get_error(), errbuf, 512);
-            throw MakeStringException(-1, "error loading certificate chain file %s - %s", certfile, errbuf);
+            throw makeStringExceptionV(-1, "error loading certificate chain file %s - %s", certFileorBuf, errbuf);
```

Comment:
could use cryptohelper::makeEVPExceptionVA instead of these 3 lines,
and similar in other places below.

Reply:
ok, changed